daemon/station_reporter.rs

```diff
@@ -0,0 +1,117 @@
+            module_name,
+            log_target,
+        }
+    }
+
+    fn print_jobs_completed(&self, total: u64) {
+        // TODO: print data from all modules
+        // https://github.com/filecoin-station/zinnia/issues/144
+        // modules: {"saturn": 100, "retrieval-checker": 23}}
+        let mut modules = Map::new();
+        modules.insert(self.module_name.clone(), json!(total));
+
+        let event = json!({
+            "type": "jobs-completed",
+            "total": total,
+             "modules": modules,
+        });
+
+        let _ = print_event(&event);
+        // ^^^ We are ignoring errors because there isn't much to do in such case
+    }
+}
+
+fn print_event(data: &serde_json::Value) -> Result<()> {
+    writeln!(stdout(), "{data}")?;
+    stdout().flush()?;
+    Ok(())
+}
+
+pub fn log_info_activity(msg: &str) {
+    let event = json!({
+        "type": "activity:info",
+        "module": serde_json::Value::Null,
+        "message": msg,
+    });
+    let _ = print_event(&event);
+    // ^^^ We are ignoring errors because there isn't much to do in such case
+}
+
+#[allow(unused)]
+pub fn log_error_activity(msg: &str) {
+    let event = json!({
+        "type": "activity:error",
+        "module": serde_json::Value::Null,
+        "message": msg,
+    });
+    let _ = print_event(&event);
+    // ^^^ We are ignoring errors because there isn't much to do in such case
+}
+
+impl Drop for StationReporter {
+    fn drop(&mut self) {
+        self.tracker
+            .borrow_mut()
+            .flush(|n| self.print_jobs_completed(n));
+    }
+}
+
+impl Reporter for StationReporter {
+    fn log(&self, level: LogLevel, msg: &str) {
+        // Important: Console logs already contain the final newline character
+        // We print all Console logs to stderr, because stdout is reserved for activity events
+        // We are ignoring write errors because there isn't much to do in such case
+        log::log!(target: &self.log_target, level.into(), "{}", msg.trim_end());
+        // let _ = write!(stderr(), "[{level:>5}] {msg}");
+        let _ = stderr().flush();
+    }
+
+    fn info_activity(&self, msg: &str) {
+        let event = json!({
+            "type": "activity:info",
+            "module": self.module_name,
+            "message": msg,
+        });
+        let _ = print_event(&event);
+        // ^^^ We are ignoring errors because there isn't much to do in such case
+    }
+
+    fn error_activity(&self, msg: &str) {
+        let event = json!({
+            "type": "activity:error",
+            "module": self.module_name,
+            "message": msg,
+        });
+        let _ = print_event(&event);
+        // ^^^ We are ignoring errors because there isn't much to do in such case
+    }
+
+    fn job_completed(&self) {
+        self.tracker
+            .borrow_mut()
+            .job_completed(|n| self.print_jobs_completed(n));
```

Comment:
```suggestion
        self.print_jobs_completed(self.tracker
            .borrow_mut()
            .job_completed());
```

API question: Is a closure the way to go, or could we simply return the value?

Reply:
So I want to log the message only if it's at least 500ms since we printed the message for the last time. 

I suppose I could change `job_completed()` and `flush()` function to return a tuple `(job_count: u64, should_log: bool)` and let the caller handle `should_log`.

```rs
let (job_count, should_log) = self.tracker.borrow_mut().job_completed();
if should_log { self.print_jobs_completed(job_count); }
```

Do you find this version easier to understand and work with?

Reply:
> So I want to log the message only if it's at least 500ms since we printed the message for the last time.

Meaning that if we call `reporter.job_completed` twice within 500ms, the first call will trigger calling the printer closure but the second call will not.

Reply:
Well, returning `(job_count, should_log)` would not work well. The tracker will not know whether it should reset the timer or not. It can reset the timer every time `job_completed` returns `should_log = true`, but that relies on the caller logging the message when told so. I find that brittle.

Reply:
Ah gotcha, I didn't know the closure was being called optionally. That works 👍